src/throttler.guard.ts

```diff
@@ -0,0 +1,169 @@
+   * Based on the context type different handlers will be called.
+   * @throws ThrottlerException
+   */
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    const handler = context.getHandler();
+    const classRef = context.getClass();
+
+    // Return early if the current route should be skipped.
+    if (this.reflector.getAllAndOverride<boolean>(THROTTLER_SKIP, [handler, classRef])) {
+      return true;
+    }
+
+    // Return early when we have no limit or ttl data.
+    const routeOrClassLimit = this.reflector.getAllAndOverride<number>(THROTTLER_LIMIT, [
+      handler,
+      classRef,
+    ]);
+    const routeOrClassTtl = this.reflector.getAllAndOverride<number>(THROTTLER_TTL, [
+      handler,
+      classRef,
+    ]);
+
+    // Check if specific limits are set at class or route level, otherwise use global options.
+    const limit = routeOrClassLimit || this.options.limit;
+    const ttl = routeOrClassTtl || this.options.ttl;
+
+    switch (context.getType<ContextType | 'graphql'>()) {
+      case 'http':
+        return this.httpHandler(context, limit, ttl);
+      case 'ws':
+        return this.websocketHandler(context, limit, ttl);
+      case 'graphql':
+        return this.graphqlHandler(context, limit, ttl);
+      default:
+        return true;
+    }
+  }
+
+  /**
+   * Throttles incoming HTTP requests.
+   * All the outgoing requests will contain RFC-compatible RateLimit headers.
+   * @see https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html#header-specifications
+   * @throws ThrottlerException
+   */
+  private async httpHandler(
+    context: ExecutionContext,
+    limit: number,
+    ttl: number,
+  ): Promise<boolean> {
+    const headerPrefix = 'X-RateLimit';
+
+    // Here we start to check the amount of requests being done against the ttl.
+    const req = context.switchToHttp().getRequest();
+    const res = context.switchToHttp().getResponse();
+
+    // Return early if the current user agent should be ignored.
+    if (Array.isArray(this.options.ignoreUserAgents)) {
+      for (const pattern of this.options.ignoreUserAgents) {
+        if (pattern.test(req.headers['user-agent'])) {
+          return true;
+        }
+      }
+    }
+
+    const key = this.generateKey(context, req.ip);
+    const ttls = await this.storageService.getRecord(key);
+    const nearestExpiryTime = ttls.length > 0 ? Math.ceil((ttls[0] - Date.now()) / 1000) : 0;
+
+    // Throw an error when the user reached their limit.
+    if (ttls.length >= limit) {
+      res.header('Retry-After', nearestExpiryTime);
+      throw new ThrottlerException();
+    }
+
+    res.header(`${headerPrefix}-Limit`, limit);
+    // We're about to add a record so we need to take that into account here.
+    // Otherwise the header says we have a request left when there are none.
+    res.header(`${headerPrefix}-Remaining`, Math.max(0, limit - (ttls.length + 1)));
+    res.header(`${headerPrefix}-Reset`, nearestExpiryTime);
```

Comment:
Just an idea: this could be extracted into a separate function that can be overridden with user options (so one can pass its own function instead and change what headers are applied etc)

Reply:
That could be an option, but we already abide the RFC standards so from our side it's already good. I say, let it be a problem first and then we can look into this. I doubt whether people want to adjust the headers.